futurex_openedx_extensions/helpers/tasks_utils.py

```diff
@@ -69,57 +69,119 @@ def _get_response_data(mocked_request: Any, kwargs: dict, view_instance: Any) ->
+        raise FXCodedException(
+            code=FXExceptionCodes.EXPORT_CSV_VIEW_RESPONSE_FAILURE,
+            message='CSV Export: The "count" key is missing or is not an int.',
+        )
+    return data, count
+
+
+def _paginated_response_generator(
+    fx_info: dict, view_data: dict, view_instance: Any
+) -> Generator:
+    """Generator to yield paginated responses."""
+    url = view_data['url']
+    kwargs = view_data.get('kwargs', {})
+    processed_records = 0
+    progress = 0
+    while url:
+        mocked_request = _get_mocked_request(url, fx_info)
+        response = view_instance(mocked_request, **kwargs)
+        data, total_records = _get_response_data(response)
+        processed_records += len(data)
+        progress = round(processed_records / total_records, 2) if total_records else 0
+        yield data, progress, processed_records
+        url = response.data.get('next')
+
+
+def _upload_file_to_storage(local_file_path: str, filename: str) -> str:
     """
     Upload a file to the default storage (e.g., S3).
 
     :param local_file_path: Path to the local file to upload
-    :param storage_path: Path in the storage where the file should be saved
+    :param filename: ilename for generated CSV
     :return: The path of the uploaded file
     """
+    storage_path = os.path.join(settings.FX_DATA_EXPORT_DIR_NAME, filename)
     with open(local_file_path, 'rb') as file:
         content_file = ContentFile(file.read())
         default_storage.save(storage_path, content_file)
     return storage_path
 
 
-def export_data_to_csv(url: str, view_data: dict, fx_permission_info: dict, filename: str) -> str:
+def _generate_csv_with_tracked_progress(
+    fx_task: Any, fx_permission_info: dict, view_data: dict, filename: str, view_instance: Any
+) -> str:
+    """
+    Generate response with progress and Write data to a CSV file.
+    :param fx_task: will be used to track progress
+    :param view_data: required data for mocking
+    :param fx_permission_info: contains role and permission info
+    :param filename: filename for generated CSV
+
+    :return: return default storage file path
+    """
+    page_size = view_data['page_size']
+    with tempfile.NamedTemporaryFile(mode='w', newline='', encoding='utf-8', delete=False) as tmp_file:
+        for data, progress, processed_records in _paginated_response_generator(
+            fx_permission_info, view_data, view_instance
+        ):
+            if data:
+                if processed_records <= page_size:
+                    # Write header only for page 1
+                    writer = csv.DictWriter(tmp_file, fieldnames=data[0].keys())
+                    writer.writeheader()
+                writer.writerows(data)
+                # update task progress
+                fx_task.progress = progress
+                fx_task.save()
+    storage_path = _upload_file_to_storage(tmp_file.name, filename)
+    os.remove(tmp_file.name)
+    return storage_path
```

Comment:
A small modification to ensure the temp file is being deleted. We also don't want the process to fail if the temp file can't be deleted for any reason
```suggestion
    try:
        with tempfile.NamedTemporaryFile(mode='w', newline='', encoding='utf-8', delete=False) as tmp_file:
            for data, progress, processed_records in _paginated_response_generator(
                fx_permission_info, view_data, view_instance
            ):
                if data:
                    if processed_records <= page_size:
                        # Write header only for page 1
                        writer = csv.DictWriter(tmp_file, fieldnames=data[0].keys())
                        writer.writeheader()
                    writer.writerows(data)
                    # update task progress
                    fx_task.progress = progress
                    fx_task.save()
        storage_path = _upload_file_to_storage(tmp_file.name, filename)
    finally:
        try:
            os.remove(tmp_file.name)
        except Exception:
            pass
    return storage_path
```